src/wrappers/hapi/HapiWrapper.ts

```diff
@@ -0,0 +1,195 @@
+
+        const server = wrappedFunction.apply(this, arguments);
+
+        /**
+         * Handler method for incoming requests & start instrumentation process
+         */
+        const startInstrument = (request: any) => ExecutionContextManager.runWithContext(
+            WrapperUtils.createExecContext, async function () {
+
+            ThundraLogger.debug('<HapiWrapper> Running with execution context');
+
+            const context: ExecutionContext = this;
+
+            request.hostname = request.info.hostname;
+            request.thundra = {
+                executionContext: context,
+            };
+
+            ThundraLogger.debug('<HapiWrapper> Before handling request');
+            await WrapperUtils.beforeRequest(request, request.response, plugins);
+        });
+
+        /**
+         * Handler method for outgoing response & finish instrumentation process
+         */
+        const finishInstrument = async (request: any, response: any) => {
+
+            ThundraLogger.debug('<HapiWrapper> Finish Instrumentation');
+
+            const context = request.thundra.executionContext;
+            context.response = response;
+
+            ExecutionContextManager.set(context);
+
+            if (response.isBoom) {
+                context.error = response;
+            }
+
+            ThundraLogger.debug('<HapiWrapper> After handling request');
+            await WrapperUtils.afterRequest(request, response, plugins, reporter);
+        };
+
+        /**
+         * Attach onPreHandler event of Hapi server
+         * it will start instrumentation process
+         */
+        server.ext('onPreHandler', (request: any, h: any) => {
+
+            ThundraLogger.debug(`<HapiWrapper> Instrumentation started for request: ${request}`);
+
+            startInstrument(request);
+
+            return h.continue;
+        });
+
+        /**
+         * Attach onPreResponse event of Hapi server
+         * it will finish instrumentation process
+         */
+        server.ext('onPreResponse', async (request: any, h: any) => {
+
+            ThundraLogger.debug(`<HapiWrapper> Instrumentation finished for request: ${request}`);
+
+            const response = request.response;
+            if (response.isBoom) {
+                const statusCode = response.output.statusCode;
+                if (statusCode === 404) {
+                    /**
+                     * if statusCode equels to 404 onPreHandler will not fired
+                     * so instrumentation process must be started in here
+                     */
+                    startInstrument(request);
+                }
+            }
+
+            await finishInstrument(request, response);
+
+            return h.continue;
+        });
+
+        return server;
+    };
+}
+
+/**
+ * Initiate Hapi wrapper & wrap Hapi server process
+ */
+export const init = () => {
+    ThundraLogger.debug('<HapiWrapper> Initializing ...');
```

Comment:
I think, we should have a config (`thundra.agent.trace.integrations.hapi.disable`) to disable Hapi wrapper explicitly in case of Hapi wrapper issue